crates/doc/src/shape/limits.rs

```diff
@@ -1,63 +1,180 @@
 // This module defines limits which are used to simplify complex,
 // typically inferred schema Shapes.
 use super::*;
-use json::Location;
-
-// Prune any locations in this shape that have more than the allowed fields,
-// squashing those fields into the `additionalProperties` subschema for that location.
-pub fn enforce_field_count_limits(slf: &mut Shape, loc: Location) {
-    // TODO: If we implement inference/widening of array tuple shapes
-    // then we'll need to also check that those aren't excessively large.
-    if slf.type_.overlaps(types::ARRAY) {
-        if let Some(array_shape) = slf.array.additional_items.as_mut() {
-            enforce_field_count_limits(array_shape, loc.push_item(0));
+use crate::ptr::Token;
+use itertools::Itertools;
+use std::cmp::Ordering;
+
+// This logic somewhat overlaps with [`Shape::locate_token`], but
+// here we don't care about recursion, we don't care about how the location
+// may or may not exist, and most importantly we need an &mut Shape.
+fn resolve_shape_mut(shape: &mut Shape, field: Token) -> Option<&mut Shape> {
+    match field {
+        Token::Index(idx) => shape.array.tuple.get_mut(idx),
+        Token::Property(prop_name) if prop_name == "*" => {
```

Comment:
nit: the `Token::Property` and `Token::NextIndex` cases are no longer possible, right? Can the become `unreachable!` ?

Reply:
Yup, `unreachable!`'d them

Reply:
I take that back. Consider
```yaml
type: object
additionalProperties:
  type: object
  properties:
    foo: string
    bar: integer
```
We need to be able to squash `/*/foo` and `/*/bar` into `/*`. I'm looking at what we need to do to make this work..

Reply:
Ok, I think the solution is: filter out locations like `/foo/bar/*` where the `*` or `-` is a leaf node. Then these can still be `unreachable!`, and we reduce the number of locations we walk